symengine/sets.cpp

```diff
@@ -301,4 +317,159 @@ const RCP<const UniversalSet> &UniversalSet::getInstance()
+        set_basic container;
+        std::set_union(container_.begin(), container_.end(),
+                       other.container_.begin(), other.container_.end(),
+                       std::inserter(container, container.begin()),
+                       RCPBasicKeyLess{});
+        return finiteset(container);
+    }
+    if (is_a<Interval>(*o)) {
+        set_basic container;
+        if (__eq__(*o))
+            return o;
+        for (const auto &a : container_) {
+            if (not o->contains(a))
+                container.insert(a);
+        }
+        if (not container.empty())
+            throw std::runtime_error("not implemented");
+        else
+            return o;
+        return o;
+    }
+    return (*o).set_union(rcp_from_this_cast<const Set>());
+}
+
+RCP<const Set> FiniteSet::set_intersection(const RCP<const Set> &o) const
+{
+    if (is_a<FiniteSet>(*o)) {
+        const FiniteSet &other = static_cast<const FiniteSet &>(*o);
+        set_basic container;
+        std::set_intersection(container_.begin(), container_.end(),
+                              other.container_.begin(), other.container_.end(),
+                              std::inserter(container, container.begin()),
+                              RCPBasicKeyLess{});
+        return finiteset(container);
+    }
+    if (is_a<Interval>(*o)) {
+        set_basic container;
+        for (const auto &a : container_) {
+            if (o->contains(a))
+                container.insert(a);
+        }
+        return finiteset(container);
+    }
+    return (*o).set_intersection(rcp_from_this_cast<const Set>());
+}
+
+bool FiniteSet::is_subset(const RCP<const Set> &o) const
+{
+    if (is_a<FiniteSet>(*o)) {
+        const FiniteSet &other = static_cast<const FiniteSet &>(*o);
+        return std::includes(other.container_.begin(), other.container_.end(),
+                             container_.begin(), container_.end(),
+                             RCPBasicKeyLess{});
+    }
+    if (is_a<Interval>(*o)) {
+        for (const auto &a : container_) {
+            if (not o->contains(a))
+                return false;
+        }
+        return true;
+    }
+    return (*o).is_superset(rcp_from_this_cast<const Set>());
+}
+
+bool FiniteSet::is_proper_subset(const RCP<const Set> &o) const
+{
+    if (is_a<FiniteSet>(*o)) {
+        const FiniteSet &other = static_cast<const FiniteSet &>(*o);
+        return (container_.size() < other.container_.size())
+               and std::includes(other.container_.begin(),
+                                 other.container_.end(), container_.begin(),
+                                 container_.end(), RCPBasicKeyLess{});
+    }
+    if (is_a<Interval>(*o)) {
+        return not __eq__(*o) and is_subset(o);
+    }
+    return (*o).is_proper_superset(rcp_from_this_cast<const Set>());
+}
+
+bool FiniteSet::is_superset(const RCP<const Set> &o) const
+{
+    if (is_a<Interval>(*o)) {
+        return __eq__(*o);
+    }
```

Comment:
Right. If I implement a new Set type and I don't handle `FiniteSet`, then it will call `superset` on the `FiniteSet` and it will call `subset` in the new Set type. This will lead to a stackoverflow.